### PR TITLE
Implement String labels

### PR DESCRIPTION
**Pull request by isaiahah**
As of this request's creation, the branch only implements string labels for the Viterbi method. These changes break the posterior method, so **do not merge**. This pull request is only for Eric to visualize the changes implemented.

### CODE REVIEW

File: segway/_util.py [hunk 1]
```diff
@@ -55,7 +55,7 @@
 
 DELIMITER_BED = '\t'  # whitespace or tab is allowed
 
-DTYPE_IDENTIFY = intc
+DTYPE_IDENTIFY = 'U16'
```
Comment (EricR86) [inline segway/_util.py:58]: Why are there single quotes on this type and not on the others (`intc`)?

File: segway/_util.py [hunk 2]
```diff
@@ -189,11 +189,15 @@ def get_col_index(chromosome, trackname):
 
 
 def get_label_color(label):
-    color = SCHEME[label % NUM_COLORS]
+    color = SCHEME[hash_label(label) % NUM_COLORS]
```
Comment (EricR86) [inline segway/_util.py:192]: Maybe there should be some data structure that holds some state and counts the current usage/collision count. So you can simply just cycle through the colors based on usage numbers.

File: segway/_util.py [hunk 3]
```diff
@@ -189,11 +189,15 @@ def get_col_index(chromosome, trackname):
...
 
     return ",".join(map(str, color))
 
 
+def hash_label(label):
+    return sum(ord(character) for character in label)
```
Comment (EricR86) [inline segway/_util.py:198]: Relatively minor but the hashlib functions are *probably* faster than this and it would simplify this code.

File: segway/_util.py [hunk 4]
```diff
@@ -346,10 +350,9 @@ def extract_superlabel(label):
...
-    if isinstance(label, str):
-        return int(label.split(".")[0])
-    else:
-        return label
+    if '.' in label:
+        return label.split(".")[0]
+    return label
 
 
 def find_segment_starts(data, output_label):
```
Comment (EricR86) [inline segway/_util.py:358]: Could you elaborate why this function changed at all? Why did `find_segment_starts` change based on the label type changing? Some additional comments on what the code is going below with the data arrays would be greatly appreciated as well.

File: segway/task.py [hunk 5]
```diff
@@ -315,9 +315,9 @@ def parse_viterbi(lines, do_reverse=False, output_label="seg"):
     # if output_label == "subseg" or "full", need to catch
     # subseg output
     if output_label != "seg":
-        re_seg = re.compile(r"^(seg|subseg)\((\d+)\)=(\d+)$")
+        re_seg = re.compile(r"^(seg|subseg)\(([\w\d]+)\)=([\w\d]+)$")
```
Comment (EricR86) [inline segway/task.py:318]: `\w` includes digits as well. Remove `\d` and the surrounding square brackets.

**Review by EricR86 — COMMENTED** [review]
Let me know when the comments have been addressed.

File: segway/_util.py (hunk 1 above)
Comment (isaiahah) [reply segway/_util.py:58]: `intc` is a numpy-defined type for signed integers compitable with C integers. `'U16'` is an array typestring, which specifies the datatype is up to 16 unicode characters. I haven't found a way to specify this without a typestring, but will add a comment explaining it.

File: segway/_util.py (hunk 2 above)
Comment (isaiahah) [reply segway/_util.py:192]: Can you elaborate on this? I don't understand what the "current usage/collision" count means.
Also because each process has a subsection of the genome, this structure would have to continue working even if the count is reset (a new process starts on a new section) or not all labels values are known (while the number of labels is known, their values are not passed).

File: segway/_util.py (hunk 3 above)
Comment (isaiahah) [reply segway/_util.py:198]: In lieu of other things, to do, I tested this and the custom function is faster (though it will be called so few times that it doesn't matter). 
More importantly, the best option among hashlib functions would look like `int(hashlib.shale_256(label.encode()).hexdigest(10), 16)`. I personally don't think that's much simpler, though I will add a comment to explain the hash_label function.

File: segway/_util.py (hunk 4 above)
Comment (isaiahah) [reply segway/_util.py:358]: `extract_superlabel` changed because it previously used the criteria of the label being a string to find `"superlabel.sublabel"`-formatted labels. However, all labels are now strings, so a new criteria (containing `'.'`) is needed. I haven't fully tested that works for superlabels.

`find_segment_starts` previously used `np.diff` to find where labels changed, as these are non-zero values. This cannot be applied to string arrays, so the code uses a very similar method to create a similar array of non-zero values at label changes.

Comments will be added.

File: segway/_util.py (hunk 1 above)
Comment (EricR86) [reply segway/_util.py]: To be clear - U16 is a 16 bit unicode type, not 16 unicode characters long? If it's a 16 bit unicode type - this should probably be a UTF-8 equivalent type instead since that's why Python defaults to anyway.

File: segway/_util.py (hunk 2 above)
Comment (EricR86) [reply segway/_util.py:192]: The idea is to avoid collisions by simply moving to the next available slot. E.g.
```python
max_collision_count = max([SCHEME_USAGE[key] for key in SCHEME_USAGE])
current_key = label % colors
if SCHEME_USAGE[current_key] >0 and SCHEME_USAGE[current_key] <= max_collision_count:
  # iterate through remaining table until you find an entry with <= max
  # put into "new key"
  SCHEME_USAGE[new_key] += 1 # update usage count
color = SCHEME[new_key]
```
Something along those lines but probably better thought out.

File: segway/_util.py (hunk 1 above)
Comment (isaiahah) [reply segway/_util.py]: `'U16'` is 16 unicode characters long. In typestrings for other datatypes the number denotes the number of bytes for that datatype, but unicode is an exception.

File: segway/_util.py (hunk 2 above)
Comment (isaiahah) [reply segway/_util.py:192]: I still don't understand how this code avoids collisions and don't see how this solves the coloring problem, so we should discuss it at the meeting. 

On this specific code, I don't think it ensures consistent coloring. Suppose labels A and B have the same key and there are two subprocesses: one over a region with label A over 75% of its length and label B over the last 25%, and one over a region with label B over 75% of its length and label A over the last 25%. These subprocesses effectively operate independently (as the information of labels and collisions in other regions is not shared between subprocesses), so by symmetry the first subprocess should assign A to `current_key` and B to `new_key` but the second subprocess should assign B to `current_key` and A to `new_key`. I might be misunderstanding the algorithm, so let me know if there's a workaround at the meeting.

File: segway/task.py [hunk 6]
```diff
@@ -278,7 +279,7 @@ def divide_posterior_array(posterior_code, num_frames, num_sublabels):
     provide the find_segment_starts() function with data in the same format
     as during the viterbi task.
     """
-    res = zeros((2, num_frames), DTYPE_IDENTIFY)
+    res = zeros((2, num_frames), DTYPE_POSTERIOR)
```
Comment (EricR86) [inline segway/task.py:282]: This is *probably* correct but I don't understand why this is part of this PR? If it looks like a bug unrelated to these changes it should be a separate PR or have an issue filed.

File: segway/_util.py [hunk 7]
```diff
@@ -342,14 +356,13 @@ def iter_chroms_coords(filenames, coords):
...
-    returned
+    label is either an integer superlabel, a string superlabel, or a string 
+    with a superlabel and sublabel part separated by a period. In this last case,
+    only the superlabel part is returned. Returns a string.
     """
-    if isinstance(label, str):
-        return int(label.split(".")[0])
-    else:
-        return label
+    if isinstance(label, str_) and '.' in label:
```
Comment (EricR86) [inline segway/_util.py:363]: why is there a `_` added to `str` here?

Comment (EricR86) [reply segway/_util.py:363]: What if there's a `.` in a given string label name?

**Review by EricR86 — COMMENTED** [review]
A few minor comments. Ping again when more questions or comments have been addressed.

File: segway/task.py (hunk 6 above)
Comment (isaiahah) [reply segway/task.py:282]: The Viterbi functions have been modified to use string labels, for CNVway. Posterior functions are not modified so only use numeric labels. I had to split the `DTYPE_IDENTIFY` constant (denoting Numpy `int_`) used for both Viterbi and Posterior functions into a `DTYPE_VITERBI` constant (denoting Numpy`str_`) and `DTYPE_POSTERIOR` constant (denoting the Numpy `int_`). Does that explain the change?

File: segway/_util.py (hunk 7 above)
Comment (isaiahah) [reply segway/_util.py:363]: `str_` is a the Numpy string class and is distinct from the Python `str` class. Only the Numpy class can be array elements, likely due to how Numpy arrays pre-allocate memory for items and Python strings are inherently extendable. 

The label is being passed directly from a Numpy array so should be a `str_`. However, for safety, I'll allow it to be either `str_` or `str`.

Comment (EricR86) [reply segway/_util.py]: Can you address symbols/names having a `.` in them in this case? There might be a need for better logic when subsegs and string names.

File: doc/segway.rst [hunk 8]
```diff
@@ -440,6 +440,12 @@ You can supply your own custom or modified models to Segway by using the
 uses. To learn more about using GMTK to create your own models there 
 the `GMTK documentation`_.
 
+The **train** and **annotate** modes of Segway accepts models where states 
+are given labeled names, rather than using the default GMTK numeric labels.
```
Comment (EricR86) [inline doc/segway.rst:444]: Should briefly specify roughly "how" this is achieved with a custom model. Which key/value is necessary in the model etc.

File: test/simplelabel/simplelabel.str [hunk 9]
```diff
@@ -0,0 +1,202 @@
+#include "traindir/auxiliary/segway.inc"
+
+GRAPHICAL_MODEL model_seg
```
Comment (EricR86) [inline test/simplelabel/simplelabel.str:3]: If possible, I would remove the unnecessary pre-processor conditional macros from this since this is going to be a simpletest. It would be nice to have as a quick reference in terms of whats necessary to get string labels to work.

File: test/simplelabel/touchstone/traindir/log/likelihood.tab [hunk 10]
```diff
@@ -0,0 +1,7 @@
+-8.71589076364230459e+03
```
Comment (EricR86) [inline test/simplelabel/touchstone/traindir/log/likelihood.tab:1]: These files match exactly locally and on the test machine? I'm surprised there isn't some need for some regex on the last 4+ digits here.

Comment (isaiahah) [reply test/simplelabel/touchstone/traindir/log/likelihood.tab:1]: I didn't implement this with Regex because these files match locally and on the test machine, and axisting tests do not generally use regex. While there are a few instances of Regex, most files use Regex on at most a few values. I will bring this up at the meeting, as it might be wise to switch all values to Regex.

File: test/simplelabel/simplelabel.str (hunk 9 above)
Comment (isaiahah) [reply test/simplelabel/simplelabel.str:3]: The structure file uses around a dozen pre-processor macros, so I don't think substituting this for individual define statments would make the test any simpler. While these macros could be substituted directly in the file, I don't think that's needed when all other tests are using this include statement in the structure file.

File: test/simplelabel/simplelabel.str [hunk 11]
```diff
@@ -11,82 +11,34 @@ frame: 0 {
    variable: seg {
       type: TYPE_SEG;
       conditionalparents: nil using DenseCPT("start_seg");
+      % Line below defines labels
```
Comment (EricR86) [inline test/simplelabel/simplelabel.str:14]: Remove "line below" and mention that is uses strings from the following name collection in input.master. For both comments in this .str file.

File: segway/_util.py [hunk 12]
```diff
@@ -189,11 +189,25 @@ def get_col_index(chromosome, trackname):
...
     return ",".join(map(str, color))
 
 
+def hash_label(label):
+    """
+    If label is an integer or a string containing an integer, return it as 
+    an integer.
+    Otherwise, hash the label string as the sum of the Unicode 
+    representations of its characters.
+    Used because hash() does not replicate across Python runs (multiple
```
Comment (EricR86) [inline segway/_util.py:203]: I would not put implementation reasoning in a doc string. A comment would be more appropriate.

**Review by EricR86 — COMMENTED** [review]
After the minor textual changes. Ping Michael for review.

File: segway/_util.py (hunk 12 above)
Comment (isaiahah) [reply segway/_util.py:203]: This isn't really needed given we're removing this coloring implementation in the other branch; nonetheless, I made the change.

**Comment by michaelmhoffman** [comment]
I thought we decided that determining colors based on a hash wasn't going to work well, and we were going to do it as part of a reduce step instead of in parallel?

**Comment by isaiahah** [comment]
We did agree to assign colors during the reduce step, apologies for the confusion of leaving in the hash code change. The change to assign colors in the reduce was implemented as part of the branch changing the output file (https://github.com/hoffmangroup/segway/pull/169), as that required significant changes to the reduction code.

I am planning to keep only that code when both branches are merged into main; however, if you would like me to make the color change in this branch as well for clarity, I can do that. Let me know what you decide.

**Comment by EricR86** [comment]
After minor miscommunication and misunderstanding, the changes have been reverted after the merge into master.
